File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        n = len(heights)
        l = 0
        r = n - 1
        ans = 0
        while l < r:
            while heights[l] < heights[r] and l < r:
                ans = max(ans, heights[l] * (r - l))
                l += 1
            while heights[l] > heights[r] and l < r:
                ans = max(ans, heights[r] * (r - l))
                r -= 1
            while heights[l] == heights[r] and l < r:
                ans = max(ans, heights[r] * (r - l))
                for i in range(l+1, r):
                    ans = max(ans, min(heights[i], heights[r]) * max(i - l, r - i))
                l += 1
                r -= 1
        return ans

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.rs
impl Solution {
    pub fn character_replacement(s: String, k: i32) -> i32 {
        use std::collections::HashMap;
        let s = s.chars().collect::<Vec<char>>();
        let mut l = 0;
        let mut r = 0;
        let mut most = s[0];
        let mut ans = 0;
        let mut map: HashMap<char, i32> = HashMap::new();
        map.insert(most, 1);
        loop {
            if ((r - l + 1) as i32) <= map.get(&most).unwrap() + k {
                ans = ans.max(r - l + 1);
                r += 1;
                if r >= s.len() {
                    break;
                }
                let a = map.entry(s[r as usize]).or_insert(0);
                *a += 1;
                if *a > *map.get(&most).unwrap() {
                    most = s[r as usize];
                }
            } else {
                let a = map.entry(s[l as usize]).or_insert(0);
                *a -= 1;
                l += 1;
                for (key, value) in map.iter() {
                    if value > map.get(&most).unwrap() {
                        most = *key;
                    }
                }
            }

        }
        ans as i32
    }
}
